### ISSUE TITLE
Question About Darcy Loss Discretization (Missing `h` Scaling?)

### ISSUE THREAD

**Issue by twMisc** [jhhuangchloe/DiffusionPDE#13]
Hi, thanks for the great codebase!

I was reviewing the Darcy loss implementation and noticed that the following lines:

https://github.com/jhhuangchloe/DiffusionPDE/blob/1e2bc8b9e312f3a936630a30d2f49aedabf0cea7/scripts/generate_darcy.py#L24-L25

might be missing a scaling factor of `1/h`. Specifically, I think these lines should be changed to

```python
    h = 1.0 / (a.shape[1] - 1) 
    deriv_x = torch.tensor([[-1, 0, 1]], dtype=torch.float64, device=device).view(1, 1, 1, 3) / (2 * h)
    deriv_y = torch.tensor([[-1], [0], [1]], dtype=torch.float64, device=device).view(1, 1, 3, 1) / (2 * h)
```

To verify this, I loaded a sample from the dataset and applied `get_darcy_loss`. Since the inputs come from the ground truth dataset, the PDE residual should be close to zero except around regions where the coefficient `a` is discontinuous. However, the original loss returns values around 1 almost everywhere. Below is a image that shows this behavior:

<img width="989" height="425" alt="Image" src="https://github.com/user-attachments/assets/f93d1c5f-1d61-48b3-99d7-ad7130fa930f" />

Here is the code I used to test this:

 <details> <summary>Click to expand</summary> 

  ```python
  import torch
  import torch.nn.functional as F
  import scipy
  import numpy as np
  import matplotlib.pyplot as plt
  
  def get_darcy_loss(a, u, a_GT, u_GT, a_mask, u_mask, device=torch.device('cuda'), h=1):
      """Return the loss of the Darcy Flow equation and the observation loss."""
      deriv_x = torch.tensor([[-1, 0, 1]], dtype=torch.float64, device=device).view(1, 1, 1, 3) / (2 * h)
      deriv_y = torch.tensor([[-1], [0], [1]], dtype=torch.float64, device=device).view(1, 1, 3, 1) / (2 * h)
      grad_x_next_x = F.conv2d(u, deriv_x, padding=(0, 1))
      grad_x_next_y = F.conv2d(u, deriv_y, padding=(1, 0))
      grad_x_next_x = a * grad_x_next_x
      grad_x_next_y = a * grad_x_next_y
      result = F.conv2d(grad_x_next_x, deriv_x, padding=(0, 1)) + F.conv2d(grad_x_next_y, deriv_y, padding=(1, 0))
      pde_loss = result + 1
      pde_loss = pde_loss.squeeze()
      
      observation_loss_a = (a - a_GT).squeeze()
      observation_loss_a = observation_loss_a * a_mask  
      observation_loss_u = (u - u_GT).squeeze()
      observation_loss_u = observation_loss_u * u_mask
      
      return pde_loss, observation_loss_a, observation_loss_u
  
  
  data = scipy.io.loadmat('darcy_1.mat')
  device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
  a_GT = torch.tensor(data['thresh_a_data'][0:1]).to(device)
  u_GT = torch.tensor(data['thresh_p_data'][0:1]).to(device)
  pde_loss, _, _ = get_darcy_loss(a_GT, u_GT, a_GT, u_GT, torch.ones_like(a_GT), torch.ones_like(u_GT), device=device)
  pde_loss_fix, _, _ = get_darcy_loss(a_GT, u_GT, a_GT, u_GT, torch.ones_like(a_GT), torch.ones_like(u_GT), device=device, h=1/(a_GT.shape[1]-1))
  
  plt.figure(figsize=(10, 5))
  plt.subplot(1, 2, 1)
  plt.imshow(np.abs(pde_loss.cpu().numpy())[2:-2, 2:-2], cmap='viridis')
  plt.colorbar(fraction=0.046, pad=0.04)
  plt.title('pde_loss')
  plt.subplot(1, 2, 2)
  plt.imshow(np.abs(pde_loss_fix.cpu().numpy())[2:-2, 2:-2], cmap='viridis')
  plt.colorbar(fraction=0.046, pad=0.04)
  plt.title('pde_loss_fix')
  plt.tight_layout()
  plt.show()
  ```

</details> 

---------------

In contrast, the following finite difference implementation (which matches the discretization in your MATLAB code [line 19–30](https://github.com/jhhuangchloe/DiffusionPDE/blob/1e2bc8b9e312f3a936630a30d2f49aedabf0cea7/dataset_generation/static/solve_gwf.m#L19-L30)) produces a more reasonable residual:
<details> <summary>Click to expand</summary> 

```python
def FDM_Darcy_match_gwf(u, a, D=1.0):
    batchsize, H, W = u.shape
    dx = D / (H - 1)

    # Interior points
    u_center = u[:, 1:-1, 1:-1]
    u_ip1 = u[:, 2:, 1:-1]
    u_im1 = u[:, :-2, 1:-1]
    u_jp1 = u[:, 1:-1, 2:]
    u_jm1 = u[:, 1:-1, :-2]

    a_center = a[:, 1:-1, 1:-1]
    a_ip1 = a[:, 2:, 1:-1]
    a_im1 = a[:, :-2, 1:-1]
    a_jp1 = a[:, 1:-1, 2:]
    a_jm1 = a[:, 1:-1, :-2]

    # Compute averaged a at half-points
    a_x_plus  = 0.5 * (a_center + a_ip1)
    a_x_minus = 0.5 * (a_center + a_im1)
    a_y_plus  = 0.5 * (a_center + a_jp1)
    a_y_minus = 0.5 * (a_center + a_jm1)

    # Flux differences (discrete divergence)
    flux_x = a_x_plus * (u_ip1 - u_center) - a_x_minus * (u_center - u_im1)
    flux_y = a_y_plus * (u_jp1 - u_center) - a_y_minus * (u_center - u_jm1)

    Du = -(flux_x + flux_y) / dx**2
    return Du

pde_residual_fdm = FDM_Darcy_match_gwf(u_GT, a_GT)
pde_residual_fdm = pde_residual_fdm.squeeze() - 1
plt.figure(figsize=(10, 5))
plt.imshow(np.abs(pde_residual_fdm.cpu().numpy()), cmap='viridis')
plt.colorbar(fraction=0.046, pad=0.04)
plt.title('FDM_Darcy_match_gwf Result')
plt.tight_layout()
plt.show()
```

</details>

---------

Please let me know if I’m missing anything. Thanks again!
